File: models/store/StockAlert.js
const { DataTypes } = require('sequelize');
const sequelize = require('../../config/database');

const StockAlert = sequelize.define('StockAlert', {
    id: {
        type: DataTypes.UUID,
        primaryKey: true,
        defaultValue: DataTypes.UUIDV4
    },
    item_id: {
        type: DataTypes.UUID,
        allowNull: false,
        references: {
            model: 'items',
            key: 'id'
        }
    },
    batch_id: {
        type: DataTypes.UUID,
        allowNull: true,
        references: {
            model: 'item_batches',
            key: 'id'
        }
    },
    institution_id: {
        type: DataTypes.UUID,
        allowNull: true
    },
    alert_type: {
        type: DataTypes.ENUM('low_stock', 'expiry_soon', 'out_of_stock', 'over_stock'),
        allowNull: false
    },
    message: {
        type: DataTypes.TEXT,
        allowNull: true
    },
    priority: {
        type: DataTypes.ENUM('low', 'medium', 'high', 'critical'),
        allowNull: false
    },
    current_quantity: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    threshold_quantity: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    expiry_date: {
        type: DataTypes.DATE,
        allowNull: true
    },
    is_resolved: {
        type: DataTypes.BOOLEAN,
        defaultValue: false
    },
    resolved_by: {
        type: DataTypes.UUID,
        allowNull: true
    },
    resolved_date: {
        type: DataTypes.DATE,
        allowNull: true
    },
    resolution_notes: {
        type: DataTypes.TEXT,
        allowNull: true
    },
    notified: {
        type: DataTypes.BOOLEAN,
        defaultValue: false
    }
}, {
    tableName: 'stock_alerts',
    timestamps: true,
});

// Define associations
StockAlert.associate = (models) => {
    StockAlert.belongsTo(models.Item, { foreignKey: 'item_id', as: 'item' });
    StockAlert.belongsTo(models.ItemBatch, { foreignKey: 'batch_id', as: 'batch' });
};

module.exports = StockAlert;

File: store-service/app/models/stock_alert.py
"""StockAlert model."""

import uuid
from datetime import datetime

from sqlalchemy import Boolean, Column, DateTime, Enum, Integer, String, Text
from sqlalchemy.dialects.postgresql import UUID

from app.database import Base
from app.utils.types import AlertType, Priority


class StockAlert(Base):
    __tablename__ = "stock_alerts"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    item_id = Column(UUID(as_uuid=True), nullable=False)
    batch_id = Column(UUID(as_uuid=True), nullable=True)
    institution_id = Column(UUID(as_uuid=True), nullable=True)
    alert_type = Column(Enum(AlertType), nullable=False)
    message = Column(Text, nullable=True)
    priority = Column(Enum(Priority), nullable=False)
    current_quantity = Column(Integer, nullable=False)
    threshold_quantity = Column(Integer, nullable=False)
    expiry_date = Column(DateTime(timezone=True), nullable=True)
    is_resolved = Column(Boolean, default=False)
    resolved_by = Column(UUID(as_uuid=True), nullable=True)
    resolved_date = Column(DateTime(timezone=True), nullable=True)
    resolution_notes = Column(Text, nullable=True)
    notified = Column(Boolean, default=False)

    created_at = Column(DateTime(timezone=True), default=datetime.utcnow)
    updated_at = Column(DateTime(timezone=True), default=datetime.utcnow, onupdate=datetime.utcnow)
